File: 543.Diameter-of-Binary-Tree/tree.c
struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
};

int diameter_of_tree(struct TreeNode *root, int *max_diam) {
    if (!root) {
        return 0;
    }

    // O(n)
    int l = diameter_of_tree(root->left, max_diam);
    int r = diameter_of_tree(root->right, max_diam);
    int this_diam = l + r;

    if (*max_diam < this_diam) {
        *max_diam = this_diam;
    }

    return 1 + (l > r ? l : r);
}

int diameterOfBinaryTree(struct TreeNode *root) {
    if (!root) {
        return 0;
    }

    int diam = 0;

    // O(n)
    diameter_of_tree(root, &diam);
    return diam;
}

File: 297.Serialize-and-Deserialize-Binary-Tree/tree.rs
use std::{cell::RefCell, rc::Rc};
type Node = Rc<RefCell<TreeNode>>;

struct Codec {}

impl Codec {
    const EMPTY: &'static str = ".";

    fn new() -> Self {
        Self {}
    }

    fn serialize(&self, root: Option<Node>) -> String {
        root.map(|node| {
            let mut node = node.borrow_mut();
            let val = node.val.to_string();
            // O(n)
            let left = self.serialize(node.left.take());
            let right = self.serialize(node.right.take());
            format!("{val} {left} {right}")
        })
        .unwrap_or(String::from(Self::EMPTY))
    }

    fn create_node<'a, I: Iterator<Item = &'a str>>(stream: &mut I) -> Option<Node> {
        let val = stream
            .next()
            .filter(|&num| num != Self::EMPTY)?
            .parse()
            .unwrap();

        let mut tree_node = TreeNode::new(val);
        tree_node.left = Self::create_node(stream);
        tree_node.right = Self::create_node(stream);
        Some(Rc::new(RefCell::new(tree_node)))
    }

    fn deserialize(&self, data: String) -> Option<Node> {
        // O(n)
        Self::create_node(&mut data.split_whitespace())
    }
}
